state/executor.go

```diff
@@ -123,6 +128,85 @@ type BlockResult struct {
 	TotalGas uint64
 }
 
+// getStakingEventType is a helper method for getting the type of system event
+func getStakingEventType(toAddress types.Address) staking.StakingEventType {
+	switch toAddress {
+	case types.StringToAddress(system.StakingAddress):
+		return staking.StakingEvent
+	case types.StringToAddress(system.UnstakingAddress):
+		return staking.UnstakingEvent
+	default:
+		return staking.UnknownEvent
+	}
+}
+
+// constructPendingEvent is a helper method for constructing a pending event
+func constructPendingEvent(
+	toAddress types.Address,
+	fromAddress types.Address,
+	value *big.Int,
+	blockNumber uint64,
+) *staking.PendingEvent {
+	eventType := getStakingEventType(toAddress)
+
+	return &staking.PendingEvent{
+		BlockNumber: int64(blockNumber),
+		Address:     fromAddress,
+		Value:       value,
+		EventType:   eventType,
+	}
+}
+
+// cleanDiscardedEvents is a helper method for clearing discarded pending events
+func cleanDiscardedEvents(transactions []*types.Transaction, blockNumber uint64) {
+	for _, discardedTxn := range transactions {
+		// Remove a pending event for the discarded transaction
+		if system.IsSystemEvent(discardedTxn.To) {
+			pendingEvent := constructPendingEvent(
+				*discardedTxn.To,
+				discardedTxn.From,
+				discardedTxn.Value,
+				blockNumber,
+			)
+
+			if pendingEvent.EventType != staking.UnknownEvent {
+				hub := staking.GetStakingHub()
+				// Remove the event from the queue if it is present
+				hub.RemovePendingEvent(*pendingEvent)
+			}
```

Comment:
What do you think of leaving this as it was, discarding any event produced by this `discardedTxn` - that way we would discard `UnknownEvents` as well, right?
This way any `UnknownEvent` ends up in the `EventQueue` for good

Reply:
You're right 👍 
Not only here, but in `commitApprovedEvents` there is a chance for `UnknownEvent`s to be leftover.
I've added a function that tidies up the queue.

Reply:
Isn't that function obsolete if you just remove this condition here and in the `commitApprovedEvents `.
This way it looks like we add the check to skip unknown events and then clean them up afterward using more CPU than necessary.
```suggestion
			hub := staking.GetStakingHub()
			// Remove the event from the queue if it is present
			hub.RemovePendingEvent(*pendingEvent)
```

Reply:
I actually believe we don't even need to check for unknown events thoroughly.

If you take a look at the commit / clear logic in the executor, we actually check if the event is a valid system event - meaning that it can never be an `Unknown` event type:
https://github.com/0xPolygon/polygon-sdk/blob/ddb93e332143a0e7c3aa92faa56a420e63a578eb/state/executor.go#L164

and

https://github.com/0xPolygon/polygon-sdk/blob/ddb93e332143a0e7c3aa92faa56a420e63a578eb/state/executor.go#L187

Removed the useless clearing now 👍 
50f5bac